cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll

```diff
@@ -109,6 +109,8 @@ private class IteratorCrementOperator extends Operator, DataFlowFunction {
   override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
     input = iteratorInput and
     output.isReturnValue()
+    or
+    input.isParameterDeref(0) and output.isReturnValueDeref()
```

Comment:
The assumption here is that if the n-th element of an iterable structure is tainted, we should assume the n+1-th element is tainted as well (i.e. the whole structure is tainted as one)?

Reply:
That's assumption is needed for the pre-increment operators, yes. Whether or not we want that I guess is debatable. It looks like that's the behavior of the AST library.

For the post-increment (and post-decrement) operators this is saying that the value pointed to by the returned iterator is the value that was stored at the current iterator's position.

Reply:
:+1:

> For the post-increment (and post-decrement) operators this is saying that the value pointed to by the returned iterator is the value that was stored at the current iterator's position.

Good point.